### PR TITLE
Add module for machine driver common functions

### PR DESCRIPTION
**Pull request by brentlu**
Add a new module to host common functions for machine drivers like dai link initialization. Here we add functions to support Intel HDMI DAI link.

I'd like to add support to other dai links as well and use the new module to handle entire dai link array in the end.

### CODE REVIEW

File: sound/soc/intel/boards/Makefile [hunk 1]
```diff
@@ -102,3 +102,6 @@ obj-$(CONFIG_SND_SOC_INTEL_SOF_NUVOTON_COMMON) += snd-soc-intel-sof-nuvoton-comm
 
 snd-soc-intel-sof-ssp-common-objs += sof_ssp_common.o
 obj-$(CONFIG_SND_SOC_INTEL_SOF_SSP_COMMON) += snd-soc-intel-sof-ssp-common.o
+
+snd-soc-intel-sof-intel-board-objs += sof_intel_board.o
+obj-$(CONFIG_SND_SOC_INTEL_SOF_INTEL_BOARD) += snd-soc-intel-sof-intel-board.o
```
Comment (plbossart) [inline sound/soc/intel/boards/Makefile]: naming is difficult!

to me "board" is something that's used for, well, board-specific quirks or setups.

Here this is really generic helpers that are not board-specific at all.

Maybe SOF_INTEL_BOARD_HELPERS or something along those lines?

File: sound/soc/intel/boards/sof_cs42l42.c [hunk 2]
```diff
@@ -491,11 +406,11 @@ static int create_bt_offload_dai_links(struct device *dev,
 static struct snd_soc_dai_link *
 sof_card_dai_links_create(struct device *dev, enum sof_ssp_codec amp_type,
 			  int ssp_codec, int ssp_amp, int ssp_bt,
-			  int dmic_be_num, int hdmi_num)
+			  int dmic_be_num, int hdmi_num, bool idisp_codec)
 {
 	struct snd_soc_dai_link_component *cpus;
 	struct snd_soc_dai_link *links;
-	int ret, id = 0, link_seq;
+	int ret, id = 0, link_seq, i;
```
Comment (plbossart) [inline sound/soc/intel/boards/sof_cs42l42.c]: I prefer to split variables that are initialized on a different line, and really all variables on different lines.

File: sound/soc/intel/boards/sof_intel_board.h [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+	struct snd_soc_jack headset_jack;
+	struct sof_hdmi_private hdmi;
+
+	enum sof_ssp_codec codec_type;
+	enum sof_ssp_codec amp_type;
+
+	int hdmi_num;
+
+	union {
+		struct sof_rt5682_private rt5682;
```
Comment (plbossart) [inline sound/soc/intel/boards/sof_intel_board.h]: why do you need a union for a single case? do you have more coming?

**Review by plbossart — COMMENTED** [review]
I only have cosmetic comments, but the most important is that this PR will conflict with upstream, I'll merge PR #4613 and you'll have to rebase @brentlu

File: sound/soc/intel/boards/sof_intel_board.h (hunk 3 above)
Comment (brentlu) [reply sound/soc/intel/boards/sof_intel_board.h]: sof_rt5682_private is for variables specific to sof_rt5682 driver. We may have more structures for other machine driver later. So yes.

File: sound/soc/intel/boards/sof_cs42l42.c [hunk 4]
```diff
@@ -491,11 +406,12 @@ static int create_bt_offload_dai_links(struct device *dev,
 static struct snd_soc_dai_link *
 sof_card_dai_links_create(struct device *dev, enum sof_ssp_codec amp_type,
 			  int ssp_codec, int ssp_amp, int ssp_bt,
-			  int dmic_be_num, int hdmi_num)
+			  int dmic_be_num, int hdmi_num, bool idisp_codec)
 {
 	struct snd_soc_dai_link_component *cpus;
 	struct snd_soc_dai_link *links;
 	int ret, id = 0, link_seq;
+	int i;
```
Comment (plbossart) [inline sound/soc/intel/boards/sof_cs42l42.c:416]: let's use one variable per line and split line 413 while we're at it.

File: sound/soc/intel/boards/sof_nau8825.c [hunk 5]
```diff
@@ -276,12 +248,11 @@ static struct snd_soc_dai_link_component dmic_component[] = {
...
 sof_card_dai_links_create(struct device *dev, enum sof_ssp_codec amp_type,
 			  int ssp_codec, int ssp_amp, int dmic_be_num,
-			  int hdmi_num)
+			  int hdmi_num, bool idisp_codec)
 {
-	struct snd_soc_dai_link_component *idisp_components;
 	struct snd_soc_dai_link_component *cpus;
 	struct snd_soc_dai_link *links;
-	int i, id = 0;
+	int i, id = 0, ret;
```
Comment (plbossart) [inline sound/soc/intel/boards/sof_nau8825.c]: let's use one variable per line and split line 255 while we're at it.

File: sound/soc/intel/boards/sof_rt5682.c [hunk 6]
```diff
@@ -642,12 +612,11 @@ static struct snd_soc_dai_link *
...
 			  enum sof_ssp_codec amp_type, int ssp_codec,
 			  int ssp_amp, int dmic_be_num, int hdmi_num,
-			  bool idisp_codec)
+			  bool idisp_codec, bool is_legacy_cpu)
 {
-	struct snd_soc_dai_link_component *idisp_components;
 	struct snd_soc_dai_link_component *cpus;
 	struct snd_soc_dai_link *links;
-	int i, id = 0;
+	int i, id = 0, ret;
```
Comment (plbossart) [inline sound/soc/intel/boards/sof_rt5682.c]: let's use one variable per line and split line 619 while we're at it.

File: sound/soc/intel/boards/sof_ssp_amp.c [hunk 7]
```diff
@@ -147,10 +120,9 @@ sof_card_dai_links_create(struct device *dev, enum sof_ssp_codec amp_type,
 			  int ssp_codec, int dmic_be_num, int hdmi_num,
 			  bool idisp_codec)
 {
-	struct snd_soc_dai_link_component *idisp_components;
 	struct snd_soc_dai_link_component *cpus;
 	struct snd_soc_dai_link *links;
-	int i, id = 0;
+	int i, id = 0, ret;
```
Comment (plbossart) [inline sound/soc/intel/boards/sof_ssp_amp.c]: let's use one variable per line and split line 125 while we're at it.

**Review by plbossart — COMMENTED** [review]
I only have nit-picks on style, otherwise this looks good to me.

**Review by plbossart — DISMISSED** [review]
Thanks @brentlu, again nice cleanup.

**Comment by plbossart** [comment]
@bardliao @kv2019i do you mind taking a look?

**Review by bardliao — APPROVED** [review]

**Review by plbossart — APPROVED** [review]

**Pull request merged**